ui/src/app/views/workspace-list/component.tsx

```diff
@@ -0,0 +1,350 @@
+  },
+  workspaceCardFooter: {
+    display: 'flex', flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center'
+  },
+  permissionBox: {
+    color: '#FFFFFF', height: '1rem', width: '3rem', fontSize: 10, textAlign: 'center',
+    borderRadius: '0.2rem', padding: 0
+  }
+});
+
+
+const WorkspaceCardMenu: React.FunctionComponent<{
+  disabled: boolean, wp: WorkspacePermissions, onShare: Function, onDelete: Function
+}> = ({
+        disabled, wp, onShare, onDelete
+      }) => {
+  const wsPathPrefix = 'workspaces/' + wp.workspace.namespace + '/' + wp.workspace.id;
+
+  return <PopupTrigger
+      data-test-id='workspace-card-menu'
+      side='bottom'
+      closeOnClick
+      content={ <React.Fragment>
+        <MenuItem icon='copy'
+                  onClick={() => {navigate([wsPathPrefix, 'clone']); }}>
+          Duplicate
+        </MenuItem>
+        <TooltipTrigger content={<div>Requires Write Permission</div>}
+                        disabled={wp.canWrite}>
+          <MenuItem icon='pencil'
+                    onClick={() => {navigate([wsPathPrefix, 'edit']); }}
+                    disabled={!wp.canWrite}>
+            Edit
+          </MenuItem>
+        </TooltipTrigger>
+        <TooltipTrigger content={<div>Requires Owner Permission</div>}
+                        disabled={wp.isOwner}>
+          <MenuItem icon='pencil' onClick={onShare} disabled={!wp.isOwner}>
+            Share
+          </MenuItem>
+        </TooltipTrigger>
+        <TooltipTrigger content={<div>Requires Owner Permission</div>}
+                        disabled={wp.isOwner}>
+          <MenuItem icon='trash' onClick={onDelete} disabled={!wp.isOwner}>
+            Delete
+          </MenuItem>
+        </TooltipTrigger>
+      </React.Fragment>}
+  >
+    <Clickable disabled={disabled} data-test-id='workspace-menu'>
+      <ClrIcon shape='ellipsis-vertical' size={21}
+               style={{color: disabled ? '#9B9B9B' : '#216FB4', marginLeft: -9,
+                 cursor: disabled ? 'auto' : 'pointer'}}/>
+    </Clickable>
+  </PopupTrigger>;
+};
+
+export class WorkspaceCard extends React.Component<
+    {wp: WorkspacePermissions, userEmail: string, reload: Function},
+    {sharing: boolean, confirmDeleting: boolean, workspaceDeletionError: boolean,
+      bugReportOpen: boolean, bugReportError: string}> {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      bugReportOpen: false,
+      bugReportError: '',
+      sharing: false,
+      confirmDeleting: false,
+      workspaceDeletionError: false
+    };
+  }
+
+  async deleteWorkspace() {
+    const {wp} = this.props;
+    workspacesApi().deleteWorkspace(wp.workspace.namespace, wp.workspace.id).then(() => {
+      this.setState({confirmDeleting: false});
+      this.props.reload();
+    }).catch(() => {
+      this.setState({workspaceDeletionError: true});
+    });
+  }
+
+  shareWorkspace(): void {
+    this.setState({sharing: false});
+    this.props.reload();
```

Comment:
So I know that this was something I did when I converted the workspace share card, but that was because I could not think of how to communicate back to the angular parent the changes from the modal child. (And I was sick of that card, so used the way that worked easily). But I think there maybe a better way here, like instead of this being a "reload" function per se, you could have this function update the state in the parent (thereby re-rendering the page, without a full reload). Questions, don't hesitate. (And I could be wrong!)

Reply:
i think this actually is the right way to do things; for all our modals we have a piece of state in the parent that says whether the modal is open or not, and the callback function when the modal closes changes the state and reloads (bc elements of the parent page may have changed with the actions of that modal, like deleting workspaces for example).  That's how we're doing it for workspace delete so i'm okay keeping it this way

Reply:
Sounds good. I'm putting on my back burner, because I think on this page for share, it's not that necessary (view won't change at all), and it would be good if we could update the page without a full reload on delete. But not stressing it now.